### PR TITLE
handle testFixtures sourceset

### LINKED ISSUE #11144: Gradle 1.7.0.CR1 regression: Gradle project does not work with test-fixtures plugin anymore
**Describe the bug**
In #10327 I was directed to use java-test-fixtures plugin for reused test resources. I am happily using it now. Unfortunately with 1.7.0.CR1 this stopped working. Reproducer can be found here: https://github.com/asodja/quarkus-test-fixtures-reproducer.

When running tests inside IDE (with Intellij Runner) I get:
```
java.lang.RuntimeException: java.io.UncheckedIOException: Error while reading file as JAR: /Users/anzes/prototype-dev/quarkus-test-fixtures-reproducible/build/libs/quarkus-test-fixtures-reproducible-my-version.jar
	at io.quarkus.test.junit.QuarkusTestExtension.throwBootFailureException(QuarkusTestExtension.java:508)
	at io.quarkus.test.junit.QuarkusTestExtension.beforeEach(QuarkusTestExtension.java:358)
...
Caused by: java.io.FileNotFoundException: /Users/anzes/prototype-dev/quarkus-test-fixtures-reproducible/build/libs/quarkus-test-fixtures-reproducible-my-version.jar (No such file or directory)
	at java.util.zip.ZipFile.open(Native Method)
	at java.util.zip.ZipFile.<init>(ZipFile.java:225)
...
```

And when I run `gradle test` from terminal I get:
```
MyResourceTest > testTestFixtures() FAILED
    java.lang.LinkageError: loader constraint violation: loader (instance of io/quarkus/bootstrap/classloading/QuarkusClassLoader) previously initiated loading for a different type with name "org/my/group/MyService"
        at java.lang.ClassLoader.defineClass1(Native Method)
        at java.lang.ClassLoader.defineClass(ClassLoader.java:757)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:420)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:363)
        at org.my.group.MyResourceTest.testTestFixtures(MyResourceTest.java:18)
```

Of course if I do `gradle build` and then run inside IntelliJ I get the same error as with `gradle test`, since jar is present in build directory.

**Expected behavior**
Test works and passes, no exception is thrown.

**Actual behavior**
Exception is thrown, test fails.

**To Reproduce**
1. git clone git@github.com:asodja/quarkus-test-fixtures-reproducer.git
2. Open in IntellIj and set Gradle: Build and run using: IntelliJ IDEA and Run tests using: IntelliJ IDEA
3. Run MyResourceTest, it fails
4. open terminal and run `gradle test`, it also fails
5. Set version of Quarkus to 1.6.1.Final
6. run `gradle test` again, it passes

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: MacBook Pro, Darwin Kernel Version 19.5.0
 - Output of `java -version`: openjdk version "1.8.0_242"
 - Quarkus version: 1.7.0.CR1
 - Build tool: Gradle 6.5.1 
---
Off topic, thank you for implementing #8737. As I follow there is still some work to do, like model caching etc, to have speed of running maven project from Intellij. But it looks like the first big step is done. Using IntelliJ runner really makes a difference when dealing with big projects, since Gradle initialization can be slow.

### PR DESCRIPTION
**Pull request by glefloch**
This branch partially fix #11144 by handling the `test-fixtures` source set. 

but the classloader issue is still present and I don't know how to fix it. Here is the error:

`````
java.lang.LinkageError: loader constraint violation: loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @27d4a09 wants to load interface org.my.group.MyService. A different interface with the same name was previously loaded by 'app'. (org.my.group.MyService is in unnamed module of loader 'app')

at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1016)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:420)
	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:363)
	at org.my.group.MyResourceTest.testTestFixtures(MyResourceTest.java:18)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.test.junit.QuarkusTestExtension.runExtensionMethod(QuarkusTestExtension.java:699)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:606)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:212)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:208)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:137)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:71)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:135)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
`````
@aloubyansky have you an idea?

### CODE REVIEW

**Comment by gsmet** [comment]
@glefloch could you point us to how to reproduce the CL issue easily? Maybe @stuartwdouglas could have a look.

**Comment by glefloch** [comment]
Yes, for sure. you can take the reproducer from #11144 and run `./gradlew clean test`.

**Comment by glefloch** [comment]
I just added an integration tests which uses the `test-fixtures` plugin.

File: integration-tests/gradle/src/test/resources/test-fixtures-module/build.gradle [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    options.compilerArgs << '-parameters'
+}
+
+compileTestJava {
+    options.encoding = 'UTF-8'
+}
+
+java {
+    sourceCompatibility = JavaVersion.VERSION_1_8
+    targetCompatibility = JavaVersion.VERSION_1_8
```
Comment (aloubyansky) [inline integration-tests/gradle/src/test/resources/test-fixtures-module/build.gradle:35]: These will need to go.

File: integration-tests/gradle/src/test/java/io/quarkus/gradle/TestFixtureModuleTestBase.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URISyntaxException;
+
+import org.junit.jupiter.api.Test;
+
+public class TestFixtureModuleTestBase extends QuarkusGradleWrapperTestBase {
```
Comment (aloubyansky) [inline integration-tests/gradle/src/test/java/io/quarkus/gradle/TestFixtureModuleTestBase.java:11]: Given that this is an actual test class, it shouldn't have Base suffix.

**Comment by stuartwdouglas** [comment]
@aloubyansky so LinkageError is because of a discrepancy between what ends up in the AppModel and the way that QuarkusBootsrap is initialized in QuarkusTestExtension. 

I have filed https://github.com/quarkusio/quarkus/pull/11213 which makes the reproducer pass when combined with this PR, but I need to talk with @aloubyansky about it.

**Comment by glefloch** [comment]
@stuartwdouglas thanks for the branch. I rebase my branch, apply @aloubyansky comments and squash my commits. 

I will start the build as soon as the fix is ok.

**Comment by stuartwdouglas** [comment]
There are still some problems with my branch, I need to talk to @aloubyansky  about it, so maybe hold off a bit.

**Comment by glefloch** [comment]
@aloubyansky @stuartwdouglas should we merge @stuartwdouglas branch or apply @aloubyansky fix?

**Comment by aloubyansky** [comment]
I'd probably go with my "fix" at this point and work on a proper one for 1.8. Unless Stuart comes up with one for 1.7.

File: test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java [hunk 3]
```diff
@@ -130,6 +131,10 @@ private ExtensionState doJavaStart(ExtensionContext context, Class<? extends Qua
                     rootBuilder.add(testResourcesLocation);
                 }
             }
+            if (Files.exists(testClassLocation.getParent().resolve("testFixtures"))) {
+                rootBuilder.add(testClassLocation.getParent().resolve("testFixtures"));
+            }
```
Comment (aloubyansky) [inline test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java:136]: @glefloch was your testcase failing w/o this change? I did try the reproducer attached by the user to verify. But I wasn't sure whether your testcase was also failing.

Comment (glefloch) [reply test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java:136]: Yes it was failing as well

**Review by aloubyansky — APPROVED** [review]

**Pull request merged**